### PR TITLE
section

### LINKED ISSUE #1326: rewrite section using page operations

### PR DESCRIPTION
**Pull request by abuts**
This should fix #1326. Section rewritten and common filebacked pattern for section is implemented

### CODE REVIEW

**Review by oerc0122 — CHANGES_REQUESTED** [review]

File: horace_core/sqw/@sqw/sqw.m [hunk 1]
```diff
@@ -116,8 +116,8 @@
         % apply alignment to pixels
         [obj,al_info] = finalize_alignment(obj,filename);
 
-        % take part of the
-        wout = section (win,varargin);
+        % take part of the object limited by fraction of the image grit
+        [wout,varargout] = section (win,varargin);
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/@sqw/sqw.m:120]: probably can be changed to irange now

File: horace_core/sqw/page_operations/PageOp_func_eval.m [hunk 2]
```diff
@@ -16,7 +16,7 @@
             %
         end
 
-        function [npix_chunks, npix_idx] = split_into_pages(~,npix,chunk_size)
+        function [npix_chunks, npix_idx,obj] = split_into_pages(obj,npix,chunk_size)
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/page_operations/PageOp_func_eval.m:19]: that how they are build. There is main implementation and there are overloads where relevant, which contain small changes. 
Unfortunately I have to change the signature of the main function, so need to change the signatures of overloads.

File: horace_core/sqw/page_operations/PageOp_section.m [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            %
+            % See split procedure for more details
+            npix_chunks = split_data_blocks(obj.block_starts_,obj.block_sizes_, chunk_size);
+            npix_idx = ones(2,numel(npix_chunks));
+            obj.block_chunks_ = npix_chunks;
+        end
+
+        function obj = apply_op(obj,varargin)
+            % Notning happens here. Selected pixels blocks are
+            % transferred to target without modifications
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/page_operations/PageOp_section.m:66]: `PageOp` takes source pixels, modifies them  and transfer them to target. What exactly the transfer means depends on if these are filebacked or memory backed pixels. Filebacked are saved, and mem-based just asigned.
What other suggestions of expressing that?

File: horace_core/sqw/page_operations/PageOp_section.m [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        end
+        function [out_obj,obj] = finish_op(obj,out_obj)
+            if numel(obj.unique_run_id_) == out_obj.experiment_info.n_runs
+                obj.check_runid = false; % this will not write experiment info again
+            else
+                % it always have to be less or equal, but some tests do not
+                % have consistent Experiment
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/page_operations/PageOp_section.m:73]: We do not request `sqw` object to be consistent or rather do not do the checks of this consistency. If you want to change this, we may discuss drawback and benifits and may introduce some or all checks, but doint this here is far out of the scope of the ticket. So, I am not fixing it here.

File: horace_core/sqw/page_operations/PageOp_section.m [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+            i = 1:size(ind_range,2);
+            sel_ranges = arrayfun(@(ii)ind_range(1,ii):ind_range(2,ii),i,'UniformOutput',false);
+            sel_blocks = block_sizes(sel_ranges{:});
+            sel_pos    = pix_start(sel_ranges{:});
+            sel_ends   = block_ends(sel_ranges{:});
+            non_empty  = sel_blocks(:)>0;
+
+            block_starts = sel_pos(non_empty);
+            block_ends   = sel_ends(non_empty);
+            % compress adjusent elements
```
[the remark this answers is not in the document]
Comment (abuts) [reply horace_core/sqw/page_operations/PageOp_section.m:109]: Yes it does exactly what `merge_ranges` does, but probably for different type of data, as `merge_ranges` seems doing a bit more, with less convenient data and looks much less efficient.

**Comment by abuts** [comment]
I am merging this as nothing serious left here and all remaining requests may be addressed in #1353

**Pull request merged**